### PR TITLE
fix: should not delete resources that are being used in a site

### PR DESCRIPTION
**Pull request by HussainTaj-arbisoft**
#### Pre-Flight checklist

- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/1754

#### What's this PR do?

When a drive_file is deleted, its related resource is also a candidate for deletion. However, if the resource is being used in some site content, like a page, deleting the resource will cause problems.

Earlier, we were only checking for resource references in the type `page` content. This PR adds additional content types to be checked, ~including `page`, `resource-list`, `video-gallery`, `navmenu`, `sitemetadata`, `course-collection`, `resource_collection`, `promo`, and `stories`~. (Edit: Now the types of content that need to be checked are fetched from site configs).

#### Where should the reviewer start?
Knowing all the different places where resources are referenced is a good idea. An excellent way to find that out is in the comment https://github.com/mitodl/ocw-studio/issues/1754#issuecomment-1515291646

`gdrive_sync/api.py::_get_content_dependencies` contains most of the modified logic.

#### Any background context you want to provide?
https://github.com/mitodl/ocw-studio/pull/1724

#### How should this be manually tested?

You will be testing two different types of sites.

##### Course

1. Download the course folder from [this drive folder](https://drive.google.com/drive/folders/1bpfyUI3bx7NuZr-G_v3ZT8Tbh9gguRBE?usp=sharing). These files will be used in this test.
2. Navigate to your local ocw-studio directory.
3. Checkout branch `hussaintaj/1754-resource-reference`
4. Run OCW Studio.
5. Create a new site with course starter.
6. Open the files downloaded in step 1. Head to the course's GDrive folder, place the content of the images directory inside `files_final` and place the content of the videos directory in `videos_final`.
7. Head back to the OCW Studio and run `Sync w/ Google Drive` for your test course.
8. **Expected Behavior**: You should see **10** resources listed on the screen after the sync is complete.
9. Publish your test course to live.  **Expected Behavior**: Publish successful.
10. Now you need to create the following content.
    * Create a page. Fill in the required fields and embed and/or link the image `page-content.jpg` in the body.
    * Create a resource list. Fill in the required fields. Add the images `resource-list-a.jpg`, and `resource-list-b.jpg` to it.
    * Create a video gallery. Fill in the required fields. Add the video `video-gallery.mp4` to it. And embed/link `video-gallery-description.jpg` to the video gallery's description.
    * Open any resource and in the description link the resource `resource-description.jpg`.
    * Navigate to the Metadata settings of the course, fill in the required content, and assign `course-image.jpg` as the course image and `course-thumbnail.jpg` as the course thumbnail.
    * Navigate to the Menu settings of the course. Add an item that links internally to the file `navmenu.jpg`.
    * Open your ocw-www site. Open or create a resource collection. Add the file `resource-collection.jpg` from your test course to this collection. Note that you must publish your test course live for the resources to appear in the list.
11. Head to the course's GDrive folder. Delete all of the image files and videos.
12. Head back to OCW Studio and run `Sync w/ Google Drive` for your test course.
13. **Expected Behavior**: None of the resources should be deleted after a successful sync.
14. **Expected Behavior**: Open the error panel (shown in the screenshot below). You should find 10 error messages, for each file we could not delete. The order is nondeterministic. 
15. Now delete everything you created in step 10. When you can't delete an item, just remove the reference of the file. You might need to use the admin panel and edit the `sitemetadata` for your test course and change `course_image_thumbnail` and `course_image` to `""`.
    - You may choose to do steps 15 and 16 one reference at a time. Each time you sync, files whose references were removed should be deleted successfully. 
16. Run `Sync w/ Google Drive` for your test course again.
17. **Expected Behavior**: All of the resources should now be deleted successfully.
18. Publish your test course.
19. **Expected Behavior**: Publishing should succeed.

##### www

1. Download the www folder from [this folder](https://drive.google.com/drive/folders/1bpfyUI3bx7NuZr-G_v3ZT8Tbh9gguRBE?usp=sharing). These files will be used in this test.
2. Navigate to your local ocw-studio directory.
3. Checkout branch `hussaintaj/1754-resource-reference`
4. Run OCW Studio.
5. Create a new site with www starter.
6. Open the files downloaded in step 1. Head to the course's GDrive folder, and place the images inside the `files_final` directory.
7. Head back to the OCW Studio and run `Sync w/ Google Drive` for your test www site.
8. **Expected Behavior**: You should see **5** resources listed on the screen after the sync is complete.
9. Publish your test www to live. **Expected Behavior**: Publish successful.
10. Now you need to create the following content.
    * Create a course collection and use `course-collection.jpg` as the image.
    * Create a promo and use `promo.jpg` as the image.
    * Create a story and use `stories.jpg` as the image.
    * Navigate to the Menu settings of the site. Add an item that links internally to the file `navmenu.jpg`.
    * Create a page. Fill in the required fields and embed and/or link the image `page-content.jpg` in the body.
11. Head to the site's GDrive folder. Delete all of the image files created in step 6.
    * You may choose to do steps 11 and 12 one reference at a time. Each time a resources should be deleted whose reference was removed.
12. Head back to OCW Studio and run `Sync w/ Google Drive` for your test course.
13. **Expected Behavior**: None of the resources should be deleted after a successful sync operation.
14. **Expected Behavior**: Open the errors panel (shown in the screenshot below). You should find 5 error messages, for each file we could not delete. The order is nondeterministic. 
15. Now delete everything you created in step 10. You can delete these items from the admin site.
16. Run `Sync w/ Google Drive` for your test site again.
17. **Expected Behavior**: All of the resources should now be deleted successfully.
18. Publish your test course.
19. **Expected Behavior**: Publishing should succeed.

#### References

> The error messages appear in the status panel thingie.
Screenshot 2023-05-11 at 1 22 01 PM

It can be a bit tricky to review the dynamic query generation. Here are the queries generated on my PC.

For course

``` sql
SELECT "websites_websitecontent"."id", "websites_websitecontent"."deleted", "websites_websitecontent"."created_on", "websites_websitecontent"."updated_on", "websites_websitecontent"."owner_id", "websites_websitecontent"."updated_by_id", "websites_websitecontent"."website_id", "websites_websitecontent"."text_id", "websites_websitecontent"."title", "websites_websitecontent"."type", "websites_websitecontent"."parent_id", "websites_websitecontent"."markdown", "websites_websitecontent"."metadata", "websites_websitecontent"."is_page_content", "websites_websitecontent"."filename", "websites_websitecontent"."dirpath", "websites_websitecontent"."file" FROM "websites_websitecontent" WHERE (("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = page AND UPPER("websites_websitecontent"."markdown"::text) LIKE UPPER(%675dcb6a-1f82-4cb5-94e2-797f1e2539c7%)) OR ("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = resource-list AND ("websites_websitecontent"."metadata" #> ['resources', 'content']) @> "675dcb6a-1f82-4cb5-94e2-797f1e2539c7") OR ("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = video_gallery AND UPPER("websites_websitecontent"."markdown"::text) LIKE UPPER(%675dcb6a-1f82-4cb5-94e2-797f1e2539c7%)) OR ("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = video_gallery AND ("websites_websitecontent"."metadata" #> ['videos', 'content']) @> "675dcb6a-1f82-4cb5-94e2-797f1e2539c7") OR ("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = resource AND UPPER("websites_websitecontent"."markdown"::text) LIKE UPPER(%675dcb6a-1f82-4cb5-94e2-797f1e2539c7%)) OR ("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = sitemetadata AND ("websites_websitecontent"."metadata" #> ['course_image', 'content']) = "675dcb6a-1f82-4cb5-94e2-797f1e2539c7") OR ("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = sitemetadata AND ("websites_websitecontent"."metadata" #> ['course_image_thumbnail', 'content']) = "675dcb6a-1f82-4cb5-94e2-797f1e2539c7") OR ("websites_websitecontent"."website_id" = a31130e7-5cb0-4859-b236-370f0302d9db AND "websites_websitecontent"."type" = navmenu AND ("websites_websitecontent"."metadata" -> leftnav) @> [{"identifier": "675dcb6a-1f82-4cb5-94e2-797f1e2539c7"}]) OR ("websites_websitecontent"."type" = resource_collections AND ("websites_websitecontent"."metadata" #> ['resources', 'content']) = [["675dcb6a-1f82-4cb5-94e2-797f1e2539c7", "courses/33-test33-spring-2023"]]))
```

For www

```sql
SELECT "websites_websitecontent"."id", "websites_websitecontent"."deleted", "websites_websitecontent"."created_on", "websites_websitecontent"."updated_on", "websites_websitecontent"."owner_id", "websites_websitecontent"."updated_by_id", "websites_websitecontent"."website_id", "websites_websitecontent"."text_id", "websites_websitecontent"."title", "websites_websitecontent"."type", "websites_websitecontent"."parent_id", "websites_websitecontent"."markdown", "websites_websitecontent"."metadata", "websites_websitecontent"."is_page_content", "websites_websitecontent"."filename", "websites_websitecontent"."dirpath", "websites_websitecontent"."file" FROM "websites_websitecontent" WHERE (("websites_websitecontent"."website_id" = 3348db73-a1d7-463e-98a2-6149c3df3620 AND "websites_websitecontent"."type" = page AND UPPER("websites_websitecontent"."markdown"::text) LIKE UPPER(%0998381f-278b-4b29-938f-ac1b80169446%)) OR ("websites_websitecontent"."type" = resource_collections AND ("websites_websitecontent"."metadata" #> ['resources', 'content']) = [["0998381f-278b-4b29-938f-ac1b80169446", null]]) OR ("websites_websitecontent"."website_id" = 3348db73-a1d7-463e-98a2-6149c3df3620 AND "websites_websitecontent"."type" = course-collection AND ("websites_websitecontent"."metadata" #> ['cover-image', 'content']) = "0998381f-278b-4b29-938f-ac1b80169446") OR ("websites_websitecontent"."website_id" = 3348db73-a1d7-463e-98a2-6149c3df3620 AND "websites_websitecontent"."type" = promos AND ("websites_websitecontent"."metadata" #> ['image', 'content']) = "0998381f-278b-4b29-938f-ac1b80169446") OR ("websites_websitecontent"."website_id" = 3348db73-a1d7-463e-98a2-6149c3df3620 AND "websites_websitecontent"."type" = stories AND ("websites_websitecontent"."metadata" #> ['image', 'content']) = "0998381f-278b-4b29-938f-ac1b80169446") OR ("websites_websitecontent"."website_id" = 3348db73-a1d7-463e-98a2-6149c3df3620 AND "websites_websitecontent"."type" = navmenu AND ("websites_websitecontent"."metadata" -> mainmenu) @> [{"identifier": "0998381f-278b-4b29-938f-ac1b80169446"}]))
```

### CODE REVIEW

**Review by gumaerc — COMMENTED** [review]
I tested this and everything worked as expected. I do have one concern, although I'm not sure it should block merging this PR. This functions by finding content using a hard coded list of lookups. The `WebsiteStarter` objects defined in https://github.com/mitodl/ocw-hugo-projects are dynamic. Content types can be added and removed with PR's against that repo.

If we accomplish this this way, then these lookups will need to be updated whenever the `WebsiteStarter` configuration changes. Ideally, this should function in a more dynamic way by looking at the starter for all content that can be linked with a relation widget, as well as markdown fields where resources can be embedded.

I tested removing one of the referenced fields from the starter and repeating the sync process and everything still worked as expected. So, we can merge this as-is, or if you think you can re-work it to instead generate the lookups dynamically based on the starter, we could do that and potentially prevent some future headaches. I think this would involve:

 - Scan the `WebsiteStarter` object for:
   - Fields with `widget: relation`
   - Fileds with `widget: menu`
   - Fields with `widget: markdown` and values in `link` and / or `embed`
 - For each of the fields with `widget: relation`, generate the lookup with something like `"__".join("metadata", field_name, "content")`
 - For each of the fields with `widget: menu`, generate a lookup based on the `name` property: `"__".join("metadata", field_name, "content"`
 - For each of the fields with `widget: markdown, add the `markdown_icontains` lookup, as this will be used with every markdown field

@pdpinch I'm interested to know your thoughts on this too. Is this out of scope for this PR? We've worked pretty hard to keep specific references to the `WebsiteStarter` config out of the code, and with the way this is it functions but has hard coded references to the way the `ocw-studio.yaml` configs are defined in `ocw-hugo-projects`. Generating the lookups dynamically would be a bit more work, but I think we should at least try to accomplish it that way unless there is an immediate need for this functionality.

**Comment by pdpinch** [comment]
Thanks for bringing this up Carey. We definitely want this code to be flexible and work with any starter. 

We can wait a bit longer for this PR.

**Comment by HussainTaj-arbisoft** [comment]
@gumaerc thank you for the review and insights. I appreciate your guidance on how to approach this matter. I'll work on making these lookups dynamic.

**Comment by HussainTaj-arbisoft** [comment]
@gumaerc I've updated the testing instructions. More specifically, I've added a few new file references to check and we no longer need to check the logs for errors.

File: gdrive_sync/api.py
Comment (gumaerc) [inline gdrive_sync/api.py:585]: I believe this could just be a function on `ConfigItem` directly called `iterate_fields` that takes an optional boolean argument of `only_cross_site`. We could also just put this in `utils.py`, but either way I tihnk `only_cross_site` should be optional and default to `False`.

File: gdrive_sync/api.py
Comment (gumaerc) [inline gdrive_sync/api.py:603]: Again, I think this could be a function directly on the `ConfigField` object.

File: gdrive_sync/api.py
Comment (gumaerc) [inline gdrive_sync/api.py:639]: This could be a function right on the `DriveFile` model.

**Review by gumaerc — CHANGES_REQUESTED** [review]
Good job making this dynamic based on the config, this is basically what I expected it would be like. I wrote a lot of feedback below but it's all related to code organization and boils down to me thinking we should either:

 - Move utility functions directly onto models
 - Move utility functions into `utils.py`

I did notice one weird situation when testing. While testing the course site on the step "Open any resource and in the description link the resource resource-description.jpg," I placed this on the description for the video. When running the sync at the end to remove the resources after removing them from GDrive, it left the resource description image saying that it was still referenced by the video description. Clicking sync a second time removed the image. It seems like maybe there's some kind of race condition where resources that are pending delete are still considered to be referencing resources that are "to be deleted." It did let me delete the resource after a second click though, so I'm not sure that it's a serious problem.

File: gdrive_sync/api.py
Comment (HussainTaj-arbisoft) [reply gdrive_sync/api.py:585]: You're right, thanks. This function is generic enough to be a part of CofigItem.

**Comment by HussainTaj-arbisoft** [comment]
@gumaerc Thank you for the review.

> I did notice one weird situation when testing. ... so I'm not sure that it's a serious problem.

This is indeed weird. Did you remove the link to `resource-description.jpg` from the video resource in step 15?

Here's what I think might have happened if the link was not removed prior to the sync.

1. `resource-description.jpg` was used by the video resource while the video was being used by nothing.
2. On sync, the async deletion task for `resource-description.jpg` ran first. Since it was being used by the video, it was not deleted.
3. Then later the deletion task for video resource ran, since nothing was using it, it was deleted.

Hence after the first sync, `resource-description.jpg` was not deleted but the video was deleted. Now because the video is gone, the reference (link) to `resource-description.jpg` is gone as well. A second sync will delete this resource.

If this is what happened then it is all expected and not a serious problem (though it is a problem in terms of UX). And I think this situation will only appear when a resource is used in another resource.

File: gdrive_sync/api.py
Comment (HussainTaj-arbisoft) [reply gdrive_sync/api.py:603]: For this, I think it would make more sense for this function to be in a `utils` module. This function only returns for fields that associate with `resource` in some way. The implementation is quite specific too. It will be returning None for the majority of the fields. Additionally, I don't think this is something we will be using quite often.

**Comment by gumaerc** [comment]
@HussainTaj-arbisoft Your explanation makes sense. I think you're right that this is a separate UX issue that we can address later and it shouldn't block this PR, but I thought I should mention it. I was able to get everything removed, just had to click the sync button twice.

As far as the code organization suggestions, your replies above make sense as well. In general I just thought that it didn't make sense for all these functions to live in `api.py`.

**Pull request merged**
